### PR TITLE
Feature/vijay register with sms

### PR DESCRIPTION
**Pull request by mob-dev**
- Created Register with SMS module
**NOTE**
- The UI flow is complete

- I have used +91 country code to verify the SMS registeration
- Also, Firebase provides login with phone via SMS. In order to authenticate users with phone number & password, we need to implement custom authentication.
- There is no option to change country code as of now
- The timer for 60 seconds before sms gets invalid part is pending (Only UI is pending, programmatically it is managed).
- The resend button to send a new sms is pending

┆Issue is synchronized with this [Jira Task](https://highvibe.atlassian.net/browse/HVN1-157) by [Unito](https://www.unito.io/learn-more)

### CODE REVIEW

File: lib/modules/auth/register/register_with_sms/verify_otp/verify_otp_controller.dart [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+final FirebaseAuth _firebaseAuth = FirebaseAuth.instance;
+
+class VerifyOtpController = _VerifyOtpControllerBase with _$VerifyOtpController;
+
+abstract class _VerifyOtpControllerBase with Store {
+  final AuthService auth = Modular.get<AuthService>();
+  final FirestoreService firestore = Modular.get<FirestoreService>();
+  final AppController appStore = Modular.get<AppController>();
+  final GlobalKey<ScaffoldState> scaffoldKey = GlobalKey<ScaffoldState>();
+  Timer timer;
```
Comment (AlexVegner) [inline lib/modules/auth/register/register_with_sms/verify_otp/verify_otp_controller.dart:24]: Make it private if you are not going to use it on UI

File: lib/modules/auth/register/register_with_sms/verify_otp/verify_otp.dart [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+  @override
+  void initState() {
+    controller.startCountDown();
+    controller.registerUser(widget.phoneNumber, widget.countryCode);
+    super.initState();
+  }
+
+  @override
+  void dispose() {
+    controller.timer.cancel();
```
Comment (AlexVegner) [inline lib/modules/auth/register/register_with_sms/verify_otp/verify_otp.dart:37]: Instantiate and dispose objects in object holder. In this case it should be disposed in controller.

File: lib/modules/auth/auth_module.dart [hunk 3]
```diff
@@ -23,6 +32,10 @@ class AuthModule extends ChildModule {
         Bind((i) => AutoLoginController()),
         Bind((i) => LoginController()),
         Bind((i) => RegisterController()),
+        Bind((i) => RegisterWithSmsController()),
```
Comment (AlexVegner) [inline lib/modules/auth/auth_module.dart:35]: Are you sure that you need lazy singleton here? I believe it's better to use factory here `, singleton: false`

File: lib/modules/auth/register/register_with_sms/register_with_sms_controller.dart [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+      final authResult = await _firebaseAuth.signInWithCredential(credential);
+      final firebaseUser = authResult.user;
+
+      if (firebaseUser.uid != null) {
+        debugPrint('googlogin userName: ${firebaseUser.uid}');
+        var userSnapshot =
+            await firestore.userCollection.document(firebaseUser.uid).get();
+        if (!userSnapshot.exists) {
+          // set user data in firebase
```
Comment (AlexVegner) [inline lib/modules/auth/register/register_with_sms/register_with_sms_controller.dart:122]: Avoid to use obvious comments

File: lib/modules/auth/register/register_with_sms/verify_otp/verify_otp.dart [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+import 'package:flutter_modular/flutter_modular.dart';
+import 'package:flutter_svg/flutter_svg.dart';
+import 'package:highvibe/utils/utils.dart';
+import 'package:highvibe/values/Strings.dart';
+import 'package:highvibe/values/themes.dart';
+import 'package:highvibe/widgets/gradient_rounded_button.dart';
+import 'package:highvibe/modules/auth/register/register_with_sms/verify_otp/verify_otp_controller.dart';
+
+class VerifyOtpPage extends StatefulWidget {
+  const VerifyOtpPage({
```
Comment (AlexVegner) [inline lib/modules/auth/register/register_with_sms/verify_otp/verify_otp.dart:12]: Avoid to use `const` for StatefulWidget widgets

**Comment by JasonHighVibe** [comment]
> * Created Register with SMS module
>   **NOTE**
> * The UI flow is complete
> * I have used +91 country code to verify the SMS registeration
> * Also, Firebase provides login with phone via SMS. In order to authenticate users with phone number & password, we need to implement custom authentication.
> * There is no option to change country code as of now
> * The timer for 60 seconds before sms gets invalid part is pending (Only UI is pending, programmatically it is managed).
> * The resend button to send a new sms is pending
> 
> ┆Issue is synchronized with this [Jira Task](https://highvibe.atlassian.net/browse/HVN1-157) by [Unito](https://www.unito.io/learn-more)

We will have to create a new flow for the SMS verification and Password setup. It will flow like this:

A user signs up with their Mobile and a OTP is generated and sent via SMS. Once they have entered the OTP, they will be taken to a password setup page where they can set-up their password. From then on, they can login via the password they have linked with their phone number.

File: lib/utils/utils.dart [hunk 6]
```diff
@@ -106,3 +107,24 @@ Future<bool> onWillPop() async {
...
 }
+
+Future<String> createProfileLink(String userId) async {
+  final parameters = DynamicLinkParameters(
+    uriPrefix: 'https://highvibe.page.link',
+    link: Uri.parse('https://highvibe.page.link/$userId'),
+    androidParameters: AndroidParameters(
+      packageName: 'network.highvibe.hvn',
+    ),
+    // iosParameters: IosParameters(
```
[the remark this answers is not in the document]
Comment (mob-dev) [reply lib/utils/utils.dart:118]: iosParameters are required. I haven't tested it yet so we should keep it for future.

**Comment by AlexVegner** [comment]
> > * Created Register with SMS module
> >   **NOTE**
> > * The UI flow is complete
> > * I have used +91 country code to verify the SMS registeration
> > * Also, Firebase provides login with phone via SMS. In order to authenticate users with phone number & password, we need to implement custom authentication.
> > * There is no option to change country code as of now
> > * The timer for 60 seconds before sms gets invalid part is pending (Only UI is pending, programmatically it is managed).
> > * The resend button to send a new sms is pending
> > 
> > ┆Issue is synchronized with this [Jira Task](https://highvibe.atlassian.net/browse/HVN1-157) by [Unito](https://www.unito.io/learn-more)
> 
> We will have to create a new flow for the SMS verification and Password setup. It will flow like this:
> 
> A user signs up with their Mobile and a OTP is generated and sent via SMS. Once they have entered the OTP, they will be taken to a password setup page where they can set-up their password. From then on, they can login via the password they have linked with their phone number.

We have 2 posible solutions for authorization method: Phone + password 
1. Use creacte own authorization methon
https://firebase.google.com/docs/auth/admin/create-custom-tokens#node.js_2 
We will need to write 3 server calls 
- Creact account with: phone + password after successful OTP authorization
Store password hash in Firebase database
```
POST /auth/v1/CreateWithPhone
Authorization: Bearer {{TOKEN}}
{
   "phone": "+91XXXXXXXXX",
   "password": "1111111",
}
```
- Authorization call 
```
POST /auth/v1/login
Authorization: Bearer {{TOKEN}}
{
   "phone": "+91XXXXXXXXX",
   "password": "1111111",
}
```
- Forgot password call (after Firebase otp)
```
POST /auth/v1/login
{
   "phone": "+91XXXXXXXXX",
   "password": "1111111",
}
```

**Pull request merged**